File: Leetcode/Python/_133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def dfs(node, mapping):
            if not node:
                return None
            new_node = Node(node.val)
            mapping[node] = new_node
            for nei in node.neighbors:
                if nei not in mapping:
                    new_nei = dfs(nei, mapping)
                new_node.neighbors.append(mapping[nei])
            return new_node
            
        return dfs(node, {})

File: Leetcode/cpp/133.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    std::unordered_map<Node*, Node*> umap; 
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return node;
        }
        if (umap.contains(node)) {
            return umap[node];
        }
        Node* add_node = new Node(node->val);;
        umap[node] = add_node;
        for (const auto& nei : node->neighbors) {
            add_node->neighbors.push_back(cloneGraph(nei));
        }
        return add_node;
    }
};
